string.c

```diff
@@ -6069,6 +6077,259 @@ rb_str_dump(VALUE str)
+	}
+	c2 = rb_enc_codepoint_len(s+1, s_end, NULL, *penc);
+	if (c2 == '{') { /* handle \u{...} form */
+	    const char *hexstr = s + 2;
+	    unsigned int hex;
+	    static const char* const close_brace = "}";
+	    long pos;
+
+	    if (hexstr >= s_end) {
+		rb_raise(rb_eArgError, "unterminated Unicode escape");
+	    }
+	    /* find close brace */
+	    pos = strseq_core(hexstr, s_end, s_end - hexstr, close_brace, 1, 0, *penc);
+	    if (pos < 0) {
+		rb_raise(rb_eArgError, "unterminated Unicode escape");
+	    }
+	    hex = ruby_scan_hex(hexstr, pos, &hexlen);
+	    if (hexlen == 0 || hexlen > 6) {
+		rb_raise(rb_eArgError, "invalid Unicode escape");
+	    }
+	    if (hex > 0x10ffffU) {
+		rb_raise(rb_eArgError, "invalid Unicode codepoint (too large)");
+	    }
+	    if ((hex & 0xfffff800U) == 0xd800U) {
+		rb_raise(rb_eArgError, "invalid Unicode codepoint");
+	    }
+	    codelen = rb_enc_codelen(hex, *penc);
+	    rb_enc_mbcput(hex, buf, *penc);
+	    rb_str_cat(undumped, buf, codelen);
+	    n += rb_strlen_lit("u{}") + hexlen;
+	}
+	else { /* handle \uXXXX form */
+	    unsigned int hex = ruby_scan_hex(s+1, 4, &hexlen);
+	    if (hexlen != 4) {
+		rb_raise(rb_eArgError, "invalid Unicode escape");
+	    }
+	    codelen = rb_enc_codelen(hex, *penc);
+	    rb_enc_mbcput(hex, buf, *penc);
+	    rb_str_cat(undumped, buf, codelen);
+	    n += rb_strlen_lit("uXXXX");
+	}
+	break;
+      case 'x':
+	if (s+1 >= s_end) {
+	    rb_raise(rb_eArgError, "invalid hex escape");
+	}
+	c2 = ruby_scan_hex(s+1, 2, &hexlen);
+	if (hexlen != 2) {
+	    rb_raise(rb_eArgError, "invalid hex escape");
+	}
+	*buf = (char)c2;
+	rb_str_cat(undumped, buf, 1L);
+	n += rb_strlen_lit("xXX");
+	break;
+      case '#':
+	if (s+1 >= s_end) {
+	    rb_str_cat(undumped, s, 1L); /* just '#' */
+	    n++;
+	    break;
+	}
+	n2 = rb_enc_mbclen(s+1, s_end, *penc);
+	if (n2 == 1 && IS_EVSTR(s+1, s_end)) {
+	    rb_str_cat(undumped, s, n);
+	    n += n2;
+	}
+	break;
+      default:
+	rb_str_cat(undumped, "\\", 1L); /* keep backslash */
+    }
+
+    return n;
+}
+
+/*
+ *  call-seq:
+ *     str.undump   -> new_str
+ *
+ *  Produces unescaped version of +str+.
+ *  See also String#dump because String#undump does inverse of String#dump.
+ *
+ *    "\"hello \\n ''\"".undump #=> "hello \n ''"
+ */
+
+static VALUE
+str_undump(VALUE str)
+{
+    const char *s = RSTRING_PTR(str);
+    const char *s_end = RSTRING_END(str);
+    long len = RSTRING_LEN(str);
+    rb_encoding *enc = rb_enc_get(str), *forced_enc;
+    int n;
+    unsigned int c;
+    enum undump_source_format source_format;
+    VALUE undumped = rb_enc_str_new(s, 0L, enc);
+    VALUE forced_enc_str;
+    long forced_enc_str_len;
+
+    rb_must_asciicompat(str);
+
+    source_format = check_undump_source_format(s, s_end, len, enc,
+					       &forced_enc_str, &forced_enc_str_len);
+    if (source_format == UNDUMP_SOURCE_INVALID) {
+	rb_raise(rb_eArgError, "not wrapped with '\"' nor '\"...\".force_encoding(\"...\")' form");
```

Comment:
Should this be ArgumentError? It seems this method has no argument.

Reply:
Thank you for comment.

> It seems this method has no argument.

Indeed.

But I couldn't find suitable one from built-in exceptions...
https://docs.ruby-lang.org/en/trunk/Exception.html

I choosed `RuntimeError` in 8db4fc678a, but better proposal is welcome.